internal/otelutil/propagation.go

```diff
@@ -0,0 +1,54 @@
+package otelutil
+
+import (
+	"context"
+	"os"
+
+	"go.opentelemetry.io/otel/propagation"
+)
+
+const (
+	traceParentKey = "traceparent"
+	traceStateKey  = "tracestate"
+
+	traceParentEnvVar = "OTEL_TRACE_PARENT"
+	traceStateEnvVar  = "OTEL_TRACE_STATE"
+)
+
+func ExtractTraceState(ctx context.Context) TraceState {
+	propagator := propagation.TraceContext{}
+	carrier := propagation.MapCarrier{}
+	propagator.Inject(ctx, carrier)
+
+	return TraceState{
+		TraceParent: carrier.Get(traceParentKey),
+		TraceState:  carrier.Get(traceStateKey),
+	}
+}
+
+type TraceState struct {
+	TraceParent string
+	TraceState  string
+}
```

Comment:
This `TraceState` is a weird mishmash of a hardcoded OTEL propagator and a hardcoded propagation carrier, with unconventional method names that don't correlate with [the terminology in the spec](https://opentelemetry.io/docs/specs/otel/context/api-propagators/).

Could we rename this to e.g. `EnvironCarrier` and have it implement `propagation.TextMapCarrier`, leaving it up to the consumer to compose it with the `TraceContext` propagator?
```go
type EnvironCarrier struct {
	TraceParent, TraceState string
}

func (c *EnvironCarrier) Get(key string) string {
	switch key {
	case traceParentKey:
		return c.TraceParent
	case traceStateKey:
		return c.TraceState
	}
	return ""
}

func (c *EnvironCarrier) Set(key, value string) {
	switch key {
	case traceParentKey:
		c.TraceParent = value
	case traceStateKey:
		c.TraceState = value
	}
	// Other keys are not supported at this time.
}

func (c *EnvironCarrier) Keys() []string {
	var k []string
	if c.TraceParent != "" {
		k = append(k, traceParentKey)
	}
	if c.TraceState != "" {
		k = append(k, traceStateKey)
	}
	return k
}

func (c *EnvironCarrier) Environ() []string {
	var env []string
	if c.TraceParent != "" {
		env = append(k, traceParentEnvVar + "=" + c.TraceParent)
	}
	if c.TraceState != "" {
		k = append(k, traceStateEnvVar + "=" + c.TraceState)
	}
	return env
}

func PropagateFromEnvironment() *EnvironCarrier {
	return &EnvironCarrier{
		TraceParent: os.Getenv(traceParentEnvVar),
		TraceState: os.Getenv(traceStateEnvVar),
	}
}
```

```go
func processSetKeyReexec() {
	tc := propagation.TraceContext{}
	otel.SetTextMapPropagator(tc) // See my review comments for setExternalKey
	carrier := otelutil.PropagateFromEnvironment()
	ctx := tc.Extract(context.Background(), carrier)

	if err := setKey(ctx); err != nil {
		_, _ = fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
}
```

```go
var carrier otelutil.EnvironCarrier
otel.GetTextMapPropagator().Inject(ctx, &carrier)
s.Hooks.Prestart = append(s.Hooks.Prestart, specs.Hook{
	// ...
	Env: carrier.Environ(),
	// ...
})
```